pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarMetadataEventSynchronizer.java

```diff
@@ -0,0 +1,199 @@
+            return;
+        }
+        this.client = (PulsarClientImpl) pulsar.getClient();
+        startProducer();
+        startConsumer();
+        log.info("Metadata event synchronizer started on topic {}", topicName);
+    }
+
+    @Override
+    public CompletableFuture<Void> notify(MetadataEvent event) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        publishAsync(event, future);
+        return future;
+    }
+
+    @Override
+    public void registerSyncListener(Function<MetadataEvent, CompletableFuture<Void>> listener) {
+        listeners.add(listener);
+    }
+
+    @Override
+    public String getClusterName() {
+        return pulsar.getConfig().getClusterName();
+    }
+
+    private void publishAsync(MetadataEvent event, CompletableFuture<Void> future) {
+        if (!started) {
+            log.info("Producer is not started on {}, failed to publish {}", topicName, event);
+            future.completeExceptionally(new IllegalStateException("producer is not started yet"));
+        }
+        producer.newMessage().value(event).sendAsync().thenAccept(__ -> {
+            log.info("successfully published metadata change event {}", event);
+            future.complete(null);
+        }).exceptionally(ex -> {
+            log.warn("failed to publish metadata update {}, will retry in {}", topicName, MESSAGE_RATE_BACKOFF_MS, ex);
+            pulsar.getBrokerService().executor().schedule(() -> publishAsync(event, future), MESSAGE_RATE_BACKOFF_MS,
+                    TimeUnit.MILLISECONDS);
+            return null;
+        });
+    }
+
+    private void startProducer() {
+        log.info("[{}] Starting producer", topicName);
+        client.newProducer(AvroSchema.of(MetadataEvent.class)).topic(topicName)
+                .messageRoutingMode(MessageRoutingMode.SinglePartition).enableBatching(false).enableBatching(false)
+                .sendTimeout(0, TimeUnit.SECONDS) //
+                .maxPendingMessages(MAX_PRODUCER_PENDING_SIZE).createAsync().thenAccept(prod -> {
+                    producer = prod;
+                    started = true;
+                    log.info("producer is created successfully {}", topicName);
+                }).exceptionally(ex -> {
+                    long waitTimeMs = backOff.next();
+                    log.warn("[{}] Failed to create producer ({}), retrying in {} s", topicName, ex.getMessage(),
+                            waitTimeMs / 1000.0);
+                    // BackOff before retrying
+                    brokerService.executor().schedule(this::startProducer, waitTimeMs, TimeUnit.MILLISECONDS);
+                    return null;
+                });
+    }
+
+    private void startConsumer() {
+        if (consumer != null) {
+            return;
+        }
+        ConsumerBuilder<MetadataEvent> consumerBuilder = client.newConsumer(AvroSchema.of(MetadataEvent.class))
```

Comment:
Nit:Please use Schema.AVRO()

Using AVRO with a Pojo that is not defined with the avro plugin leads to bad performance because it is going to use reflection.

We could generate the AVRO poco using the Maven plugin. I am 100% that it is worth (but not to be done in this patch)

Reply:
I can change to `Schema.AVRO()` but it eventually calls [AvroSchema.of](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImplementationBindingImpl.java#L208). Also, it is called only once at server startup so, it should not have a performance issue.

Reply:
as I said, it won't make much of a difference but I still have made the change so, we can progress for this PR.